okio/src/jvmTest/kotlin/okio/SocketTest.kt

```diff
@@ -0,0 +1,303 @@
+    assertThat(javaNetSocket.isOutputShutdown).isFalse()
+    assertThat(javaNetSocket.isClosed).isFalse()
+  }
+
+  @Test
+  fun closeSinkDoesNotCloseJavaNetSocket() {
+    val javaNetSocket = (this.socket as? DefaultSocket)?.socket ?: return
+
+    socket.sink.close()
+    assertThat(javaNetSocket.isInputShutdown).isFalse()
+    assertThat(javaNetSocket.isOutputShutdown).isTrue()
+    assertThat(javaNetSocket.isClosed).isFalse()
+  }
+
+  @Test
+  fun closeSourceThenSinkClosesJavaNetSocket() {
+    val javaNetSocket = (this.socket as? DefaultSocket)?.socket ?: return
+
+    socket.source.close()
+    socket.sink.close()
+    assertThat(javaNetSocket.isClosed).isTrue()
+  }
+
+  @Test
+  fun closeSinkThenSourceClosesJavaNetSocket() {
+    val javaNetSocket = (this.socket as? DefaultSocket)?.socket ?: return
+
+    socket.sink.close()
+    socket.source.close()
+    assertThat(javaNetSocket.isClosed).isTrue()
+  }
+
+  @Test
+  fun closeSinkThenSourceClosesJavaNetSocketEvenIfStreamsAlreadyClosed() {
+    val javaNetSocket = (this.socket as? DefaultSocket)?.socket ?: return
+    javaNetSocket.shutdownInput()
+    javaNetSocket.shutdownOutput()
+    assertThat(javaNetSocket.isClosed).isFalse()
+
+    socket.sink.close()
+    socket.source.close()
+    assertThat(javaNetSocket.isClosed).isTrue()
+  }
+
+  @Test
+  fun closeSourceIsIdempotent() {
+    val javaNetSocket = (this.socket as? DefaultSocket)?.socket ?: return
+
+    socket.source.close()
+    assertThat(javaNetSocket.isInputShutdown).isTrue()
+    assertThat(javaNetSocket.isClosed).isFalse()
+    socket.source.close()
+    assertThat(javaNetSocket.isInputShutdown).isTrue()
+    assertThat(javaNetSocket.isClosed).isFalse()
+  }
+
+  @Test
+  fun closeSinkIsIdempotent() {
+    val javaNetSocket = (this.socket as? DefaultSocket)?.socket ?: return
+
+    socket.sink.close()
+    assertThat(javaNetSocket.isOutputShutdown).isTrue()
+    assertThat(javaNetSocket.isClosed).isFalse()
+    socket.sink.close()
+    assertThat(javaNetSocket.isOutputShutdown).isTrue()
+    assertThat(javaNetSocket.isClosed).isFalse()
+  }
+
+  enum class Factory {
+    /** Implements an okio.Socket using the `java.net.Socket` API on OS sockets. */
+    Default {
+      override fun createSocketPair(): Pair<Socket, Socket> {
+        val localhost = InetAddress.getByName("localhost")
+
+        val serverSocket = ServerSocket()
+        serverSocket.bind(InetSocketAddress(localhost, 0))
+
+        val socketBFuture = CompletableFuture<java.net.Socket>()
+        thread(name = "createSocketPair") {
+          socketBFuture.complete(serverSocket.accept())
+        }
+
+        val socketA = SocketFactory.getDefault().createSocket()
+        socketA.connect(InetSocketAddress(localhost, serverSocket.localPort))
+
+        val socketB = socketBFuture.get()
+        return socketA.asOkioSocket() to socketB.asOkioSocket()
+      }
+    },
+
+    /** Implements an okio.Socket using a pair of in-memory pipes. */
+    Pipes {
```

Comment:
Did you also test to use this in tests?

Reply:
The tests exercise both thanks to `@Burst`.